gms/gossiper.cc

```diff
@@ -2088,7 +2088,7 @@ future<> gossiper::add_saved_endpoint(inet_address ep) {
     if (es) {
         ep_state = *es;
         logger.debug("not replacing a previous ep_state for {}, but reusing it: {}", ep, ep_state);
-        ep_state.set_heart_beat_state_and_update_timestamp(heart_beat_state());
+        ep_state.update_timestamp();
```

Comment:
We cannot reuse existing generation if we set endpoint state version using our local version generator, e.g. here:
```
 ep_state.add_application_state(gms::application_state::HOST_ID, versioned_value::host_id(host_id.value()));
```

Local node's highest version may be ahead of the ep_state node's highest version. If we gossip this state around, other nodes will ignore changes to states made by the original node. That's because gossip protocol is incremental, node's request only states with versions higher than the highest known version.

Setting generation to 0 works because it will always lose with the states gossiped by the original node.

Reply:
I've been thinking about that too actually.
Once the endpoint state is set with a non-zero generation we can assume it was set by gossip directly or indirectly from the node itself, since other nodes will add their saved info about it with generation=0, as we do (now), and so add_saved_endpoint should not update it in this case.
I don't think that was ever intended so it might even warrant an internal error.